_posts/2021-03-15-forecasting-time-series-with-torch_3/forecasting_3.Rmd

```diff
@@ -0,0 +1,445 @@
+  
+  initialize = function(x, n_timesteps, sample_frac = 1) {
+    
+    self$n_timesteps <- n_timesteps
+    self$x <- torch_tensor((x - train_mean) / train_sd)
+    
+    n <- length(self$x) - self$n_timesteps - 1
+    
+    self$starts <- sort(sample.int(
+      n = n,
+      size = n * sample_frac
+    ))
+    
+  },
+  
+  .getitem = function(i) {
+    
+    start <- self$starts[i]
+    end <- start + self$n_timesteps - 1
+    lag <- 1
+    
+    list(
+      x = self$x[start:end],
+      y = self$x[(start+lag):(end+lag)]$squeeze(2)
+    )
+    
+  },
+  
+  .length = function() {
+    length(self$starts) 
+  }
+)
+```
+
+Dataset as well as dataloader instantations then can proceed as before.
+
+```{r}
+batch_size <- 32
+
+train_ds <- elec_dataset(elec_train, n_timesteps, sample_frac = 0.5)
+train_dl <- train_ds %>% dataloader(batch_size = batch_size, shuffle = TRUE)
+
+valid_ds <- elec_dataset(elec_valid, n_timesteps, sample_frac = 0.5)
+valid_dl <- valid_ds %>% dataloader(batch_size = batch_size)
+
+test_ds <- elec_dataset(elec_test, n_timesteps)
+test_dl <- test_ds %>% dataloader(batch_size = 1)
+```
+
+# Model
+
+Technically, the model consists of three *modules*: the aforementioned encoder and decoder, and the *seq2seq* module that orchestrates them.
+
+### Encoder
+
+The encoder takes its input and runs it through an RNN. Of the two things returned by a recurrent neural network, outputs and state, so far we've only been using the former. This time, we do the opposite: We throw away the outputs, and only return the state.
+
+If the RNN in question is a GRU (and assuming that of the outputs, we take just the final time step, which is what we've been doing throughout), there really is no difference: The final state equals the final output. If it's an LSTM, however, there is a second, the "cell", state. In that case, returning the state instead of the final output will carry more information.
+
+```{r}
+encoder_module <- nn_module(
+  
+  initialize = function(type, input_size, hidden_size, num_layers = 1, dropout = 0) {
+    
+    self$type <- type
+    
+    self$rnn <- if (self$type == "gru") {
+      nn_gru(
+        input_size = input_size,
+        hidden_size = hidden_size,
+        num_layers = num_layers,
+        dropout = dropout,
+        batch_first = TRUE
+      )
+    } else {
+      nn_lstm(
+        input_size = input_size,
+        hidden_size = hidden_size,
+        num_layers = num_layers,
+        dropout = dropout,
+        batch_first = TRUE
+      )
+    }
+    
+  },
+  
+  forward = function(x) {
+    
+    x <- self$rnn(x)
+    
+    # return last states for all layers
+    # per layer, a single tensor for GRU, a list of 2 tensors for LSTM
+    x <- x[[2]]
+    x
+    
+  }
+  
+)
+```
+
+### Decoder
+
+In the decoder, too, the main component is an RNN. In contrast to previously-shown architectures, though, it does not just return a prediction. It also reports back the RNN's final state.
```

Comment:
```suggestion
In the decoder, the main component is an RNN. In contrast to previously-shown architectures, though, it does not just return a prediction. It also reports back the RNN's final state.
```

Reply:
ok, perhaps I should make more clear what the reference is, here ....as in
> In the decoder, just like in the encoder, the main component